File: frontend/lib/api/photo_api.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'api_config.dart';
import 'jwt_service.dart';
import '../models/photo.dart';
import '../services/upload_progress_service.dart';

/// Récupère les photos d'une fuite.
/// [limit] optionnel : limite le nombre de photos retournées.
Future<List<Photo>> getPhotosByFuite(int fuiteId, {int? limit}) async {
  final headers = await authHeaders();
  var url = '${ApiConfig.apiBaseUrl}/photos?fuiteId=$fuiteId';
  if (limit != null && limit > 0) url += '&limit=$limit';
  final response = await http
      .get(Uri.parse(url), headers: headers)
      .timeout(ApiConfig.timeout);

  if (response.statusCode == 200) {
    final List<dynamic> jsonList = jsonDecode(response.body);
    return jsonList
        .map((json) => Photo.fromJson(json as Map<String, dynamic>))
        .toList();
  }
  throw Exception('Erreur ${response.statusCode}: ${response.body}');
}

/// Récupère une photo par son ID.
Future<Photo> getPhotoById(int id) async {
  final headers = await authHeaders();
  final response = await http
      .get(Uri.parse('${ApiConfig.apiBaseUrl}/photos/$id'), headers: headers)
      .timeout(ApiConfig.timeout);

  if (response.statusCode == 200) {
    return Photo.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
  }
  throw Exception('Erreur ${response.statusCode}: ${response.body}');
}

/// Crée une nouvelle photo (upload multipart).
/// [thumbnailPath] optionnel : chemin local vers la miniature (pour les vidéos).
/// [onProgress] optionnel : callback appelé régulièrement avec la fraction
/// (0.0 → 1.0) envoyée. Permet d'afficher une barre de progression réelle.
Future<Photo> createPhoto({
  required String cheminFichier,
  String? datePrise,
  String? annotationsDessin,
  required int fuiteId,
  String? thumbnailPath,
  void Function(double progress)? onProgress,
}) async {
  final headers = await authHeaders();
  // On enlève Content-Type pour que http package mette multipart/form-data
  headers.remove('Content-Type');

  final request = http.MultipartRequest(
    'POST',
    Uri.parse('${ApiConfig.apiBaseUrl}/photos/upload'),
  );
  request.headers.addAll(headers);
  request.fields['fuiteId'] = fuiteId.toString();
  if (datePrise != null) request.fields['datePrise'] = datePrise;
  if (annotationsDessin != null) {
    request.fields['annotationsDessin'] = annotationsDessin;
  }

  // Fichier principal avec suivi de progression réel :
  // on compte les octets lus depuis le fichier pendant l'envoi.
  final file = File(cheminFichier);
  final totalBytes = await file.length();
  var sentBytes = 0;

  if (onProgress != null && totalBytes > 0) {
    // Stream du fichier avec comptage des octets pour la progression
    final stream = file.openRead().transform(
      StreamTransformer<List<int>, List<int>>.fromHandlers(
        handleData: (List<int> data, EventSink<List<int>> sink) {
          sentBytes += data.length;
          onProgress((sentBytes / totalBytes).clamp(0.0, 1.0).toDouble());
          sink.add(data);
        },
      ),
    );
    request.files.add(
      http.MultipartFile(
        'file',
        stream,
        totalBytes,
        filename: file.uri.pathSegments.last,
      ),
    );
  } else {
    request.files.add(await http.MultipartFile.fromPath('file', cheminFichier));
  }

  if (thumbnailPath != null) {
    request.files.add(
      await http.MultipartFile.fromPath('thumbnail', thumbnailPath),
    );
  }

  // Marquage upload en cours → le ConnectivityService ne ping pas
  UploadProgressService.instance.beginUpload();
  try {
    final streamedResponse = await request.send().timeout(
      const Duration(minutes: 15),
    );
    final response = await http.Response.fromStream(streamedResponse);
    if (response.statusCode == 201) {
      return Photo.fromJson(jsonDecode(response.body) as Map<String, dynamic>);
    }
    throw Exception('Erreur ${response.statusCode}: ${response.body}');
  } finally {
    UploadProgressService.instance.endUpload();
  }
}

/// Supprime une photo.
Future<void> deletePhoto(int id) async {
  final headers = await authHeaders();
  final response = await http
      .delete(Uri.parse('${ApiConfig.apiBaseUrl}/photos/$id'), headers: headers)
      .timeout(ApiConfig.timeout);

  if (response.statusCode != 204) {
    throw Exception('Erreur ${response.statusCode}: ${response.body}');
  }
}

File: web/src/api/photoApi.ts
import { requestMultipart, request } from './request'
import { getAuthHeaders } from './jwtService'
import { ACTIVE_API_URL } from '../config/baseURL'
import type { PhotoResponseDto } from '../types'

/**
 * Upload une photo/vidéo pour une fuite.
 * Endpoint backend : POST /api/photos/upload (multipart).
 */
export async function uploadPhoto(
  fuiteId: number,
  file: File,
  datePrise?: string,
): Promise<PhotoResponseDto> {
  const formData = new FormData()
  formData.append('file', file)
  formData.append('fuiteId', String(fuiteId))
  if (datePrise) formData.append('datePrise', datePrise)
  return requestMultipart<PhotoResponseDto>('/photos/upload', formData)
}

/**
 * Upload une photo/vidéo avec suivi de progression (XHR).
 * `onProgress` reçoit un nombre entre 0 et 1.
 * Même logique que le mobile (photo_api.dart) : comptage des octets envoyés.
 */
export async function uploadPhotoWithProgress(
  fuiteId: number,
  file: File,
  onProgress: (progress: number) => void,
  datePrise?: string,
): Promise<PhotoResponseDto> {
  const formData = new FormData()
  formData.append('file', file)
  formData.append('fuiteId', String(fuiteId))
  if (datePrise) formData.append('datePrise', datePrise)

  return new Promise<PhotoResponseDto>((resolve, reject) => {
    const xhr = new XMLHttpRequest()
    xhr.open('POST', `${ACTIVE_API_URL}/photos/upload`)
    const headers = getAuthHeaders()
    if (headers.Authorization) {
      xhr.setRequestHeader('Authorization', headers.Authorization)
    }
    xhr.timeout = 15 * 60 * 1000 // 15 min pour les vidéos volumineuses (comme le mobile)

    xhr.upload.onprogress = (e) => {
      if (e.lengthComputable && e.total > 0) {
        onProgress(Math.min(1, e.loaded / e.total))
      }
    }

    xhr.onload = () => {
      try {
        const data = JSON.parse(xhr.responseText) as PhotoResponseDto
        if (xhr.status >= 200 && xhr.status < 300) {
          resolve(data)
        } else {
          const msg =
            (data as unknown as { message?: string; error?: string })?.message ||
            (data as unknown as { error?: string })?.error ||
            `Request failed with status ${xhr.status}`
          reject(new Error(msg))
        }
      } catch {
        reject(new Error('Réponse invalide du serveur'))
      }
    }

    xhr.onerror = () => reject(new Error('Erreur réseau lors de l\'upload'))
    xhr.ontimeout = () => reject(new Error("L'upload a dépassé le délai (15 min)"))
    xhr.send(formData)
  })
}

/** Liste les photos d'une fuite (limit optionnelle, comme le mobile). */
export async function getPhotosByFuite(
  fuiteId: number,
  limit?: number,
): Promise<PhotoResponseDto[]> {
  const qs = limit && limit > 0 ? `&limit=${limit}` : ''
  return request<PhotoResponseDto[]>(`/photos?fuiteId=${fuiteId}${qs}`)
}

/** Récupère une photo par ID. */
export async function getPhotoById(id: number): Promise<PhotoResponseDto> {
  return request<PhotoResponseDto>(`/photos/${id}`)
}

/** Supprime une photo. */
export async function deletePhoto(photoId: number): Promise<void> {
  await request<unknown>(`/photos/${photoId}`, { method: 'DELETE' })
}
